Serialization of a tree

Is it possible to serialize/deserialize a tree? When I try the following two variants, I get two different errors that follows:

```python
from typing import List
from serde import serde

@dataclass
class Node:
    name: str
    children: List["Node"]

serde(Node)
```

fails with: 

```
During handling of the above exception, another exception occurred:
test_context.py:12: in <module>
    serde(Node)
../../../.cache/pypoetry/virtualenvs/querychains-RLdrqFKA-py3.10/lib/python3.10/site-packages/serde/__init__.py:153: in serde
    return wrap(_cls)
../../../.cache/pypoetry/virtualenvs/querychains-RLdrqFKA-py3.10/lib/python3.10/site-packages/serde/__init__.py:146: in wrap
    serialize(cls, **kwargs)
../../../.cache/pypoetry/virtualenvs/querychains-RLdrqFKA-py3.10/lib/python3.10/site-packages/serde/se.py:311: in serialize
    return wrap(_cls)
../../../.cache/pypoetry/virtualenvs/querychains-RLdrqFKA-py3.10/lib/python3.10/site-packages/serde/se.py:283: in wrap
    for union in iter_unions(cls):
../../../.cache/pypoetry/virtualenvs/querychains-RLdrqFKA-py3.10/lib/python3.10/site-packages/serde/compat.py:412: in iter_unions
    recursive(cls)
../../../.cache/pypoetry/virtualenvs/querychains-RLdrqFKA-py3.10/lib/python3.10/site-packages/serde/compat.py:394: in recursive
    recursive(f.type)
../../../.cache/pypoetry/virtualenvs/querychains-RLdrqFKA-py3.10/lib/python3.10/site-packages/serde/compat.py:402: in recursive
    recursive(args[0])

 ... SKIPPED MANY LINES ...

../../../.cache/pypoetry/virtualenvs/querychains-RLdrqFKA-py3.10/lib/python3.10/site-packages/serde/compat.py:393: in recursive
    for f in dataclass_fields(cls):
../../../.cache/pypoetry/virtualenvs/querychains-RLdrqFKA-py3.10/lib/python3.10/site-packages/serde/compat.py:299: in dataclass_fields
    raise SerdeError(
E   serde.compat.SerdeError: Failed to resolve type hints for Node:
E   RecursionError: maximum recursion depth exceeded while calling a Python object
E   
E   If you are using forward references make sure you are calling deserialize & serialize after all classes are globally visible.
```

If I switch from "List" to "list"

```python
@dataclass
class Node:
    name: str
    children: list["Node"]

serde(Node)
```

Fails with:

```
test_context.py:12: in <module>
    serde(Node)
../../../.cache/pypoetry/virtualenvs/querychains-RLdrqFKA-py3.10/lib/python3.10/site-packages/serde/__init__.py:153: in serde
    return wrap(_cls)
../../../.cache/pypoetry/virtualenvs/querychains-RLdrqFKA-py3.10/lib/python3.10/site-packages/serde/__init__.py:146: in wrap
    serialize(cls, **kwargs)
../../../.cache/pypoetry/virtualenvs/querychains-RLdrqFKA-py3.10/lib/python3.10/site-packages/serde/se.py:311: in serialize
    return wrap(_cls)
../../../.cache/pypoetry/virtualenvs/querychains-RLdrqFKA-py3.10/lib/python3.10/site-packages/serde/se.py:295: in wrap
    add_func(scope, TO_ITER, render_to_tuple(cls, serializer, type_check, serialize_class_var), g)
../../../.cache/pypoetry/virtualenvs/querychains-RLdrqFKA-py3.10/lib/python3.10/site-packages/serde/se.py:513: in render_to_tuple
    return env.get_template("iter").render(
../../../.cache/pypoetry/virtualenvs/querychains-RLdrqFKA-py3.10/lib/python3.10/site-packages/jinja2/environment.py:1301: in render
    self.environment.handle_exception()
../../../.cache/pypoetry/virtualenvs/querychains-RLdrqFKA-py3.10/lib/python3.10/site-packages/jinja2/environment.py:936: in handle_exception
    raise rewrite_traceback_stack(source=source)
<template>:19: in top-level template code
    ???
../../../.cache/pypoetry/virtualenvs/querychains-RLdrqFKA-py3.10/lib/python3.10/site-packages/serde/se.py:698: in render
    res = self.list(arg)
../../../.cache/pypoetry/virtualenvs/querychains-RLdrqFKA-py3.10/lib/python3.10/site-packages/serde/se.py:787: in list
    return f"[{self.render(earg)} for v in {arg.varname}]"
../../../.cache/pypoetry/virtualenvs/querychains-RLdrqFKA-py3.10/lib/python3.10/site-packages/serde/se.py:714: in render
    res = self.primitive(arg)
../../../.cache/pypoetry/virtualenvs/querychains-RLdrqFKA-py3.10/lib/python3.10/site-packages/serde/se.py:837: in primitive
    typ = typename(arg.type)
../../../.cache/pypoetry/virtualenvs/querychains-RLdrqFKA-py3.10/lib/python3.10/site-packages/serde/compat.py:256: in typename
    return typ.__name__
E   AttributeError: 'str' object has no attribute '__name__'. Did you mean: '__ne__'?
```

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -19,12 +19,3 @@ repos:
                "--skip-string-normalization",
               ]
 
-  - repo: https://github.com/PyCQA/isort
-    rev: 5.12.0
-    hooks:
-      - id: isort
-        args: [
-               "--profile=black",
-               "--line-length=120",
-               "--atomic",
-              ]

--- a/CONTRIBUTING.md
+++ b/CONTRIBUTING.md
@@ -9,7 +9,7 @@ Thank you for considering contributing to Pyserde!
 - List your Python and Pyserde versions. If possible, check if this issue is already fixed in the repository.
 
 ## Submitting patches
-- Pyserde uses Black and isort to autoformat your code. This should be done for you as a git pre-commit hook, which gets installed when you run `make setup` but you can do it manually via `make fmt`.
+- Pyserde uses Black to autoformat your code. This should be done for you as a git pre-commit hook, which gets installed when you run `make setup` but you can do it manually via `make fmt`.
 - Include tests if your patch is supposed to solve a bug, and explain clearly under which circumstances the bug happens. Make sure the test fails without your patch.
 - Include a string like “Fixes #123” in your commit message (where 123 is the issue you fixed). See [Closing issues using keywords](https://help.github.com/articles/creating-a-pull-request/).
 

--- a/examples/recursive_list.py
+++ b/examples/recursive_list.py
@@ -1,0 +1,25 @@
+from dataclasses import dataclass
+from typing import List
+
+from serde import serde
+from serde.json import from_json, to_json
+
+
+@dataclass
+class Node:
+    name: str
+    children: List["Node"]
+
+
+serde(Node)
+
+
+def main() -> None:
+    n = Node("a", [Node("b", [Node("c", [])])])
+    s = to_json(n)
+    print(f"Into Json: {s}")
+    print(f"From Json: {from_json(Node, s)}")
+
+
+if __name__ == "__main__":
+    main()

--- a/examples/runner.py
+++ b/examples/runner.py
@@ -24,6 +24,7 @@ import pep681
 import plain_dataclass
 import plain_dataclass_class_attribute
 import recursive
+import recursive_list
 import rename
 import rename_all
 import simple
@@ -78,6 +79,7 @@ def run_all():
     run(class_var)
     run(alias)
     run(recursive)
+    run(recursive_list)
     run(class_var)
     run(plain_dataclass)
     run(plain_dataclass_class_attribute)

--- a/serde/compat.py
+++ b/serde/compat.py
@@ -353,11 +353,6 @@ def iter_types(cls: TypeLike) -> List[TypeLike]:
             args = type_args(cls)
             if args:
                 recursive(args[0])
-        elif is_set(cls):
-            lst.add(Set)
-            args = type_args(cls)
-            if args:
-                recursive(args[0])
         elif is_tuple(cls):
             lst.add(Tuple)
             for arg in type_args(cls):
@@ -380,9 +375,12 @@ def iter_unions(cls: TypeLike) -> List[TypeLike]:
     Iterate over all unions that are used in the dataclass
     """
     lst: Set[TypeLike] = set()
+    stack: List[TypeLike] = []  # To prevent infinite recursion
 
     def recursive(cls: TypeLike) -> None:
         if cls in lst:
+            return
+        if cls in stack:
             return
 
         if is_union(cls):
@@ -390,8 +388,10 @@ def iter_unions(cls: TypeLike) -> List[TypeLike]:
             for arg in type_args(cls):
                 recursive(arg)
         if is_dataclass(cls):
+            stack.append(cls)
             for f in dataclass_fields(cls):
                 recursive(f.type)
+            stack.pop()
         elif is_opt(cls):
             args = type_args(cls)
             if args:

--- a/setup.cfg
+++ b/setup.cfg
@@ -6,8 +6,6 @@ addopts = -v
 max-line-length = 120
 ignore = E252,W503
 max-complexity = 30
-[isort]
-line_length = 120
 [mypy]
 ignore_missing_imports = True
 strict = True

--- a/tests/common.py
+++ b/tests/common.py
@@ -9,7 +9,6 @@ import uuid
 from typing import Any, Callable, DefaultDict, Dict, FrozenSet, Generic, List, NewType, Optional, Set, Tuple, TypeVar
 
 import more_itertools
-
 from serde import from_dict, from_tuple, serde, to_dict, to_tuple
 from serde.json import from_json, to_json
 from serde.msgpack import from_msgpack, to_msgpack
@@ -92,6 +91,11 @@ types: List = [
     param(data.Pri(10, 'foo', 100.0, True), Optional[data.Pri]),
     param(None, Optional[data.Pri], toml_not_supported),
     param(data.Recur(data.Recur(None, None, None), None, None), data.Recur, toml_not_supported),
+    param(
+        data.RecurContainer([data.RecurContainer([], {})], {"c": data.RecurContainer([], {})}),
+        data.RecurContainer,
+        toml_not_supported,
+    ),
     param(data.Init(1), data.Init),
     param(10, NewType('Int', int)),  # NewType
     param({'a': 1}, Any),  # Any

--- a/tests/data.py
+++ b/tests/data.py
@@ -236,6 +236,15 @@ class Recur:
 serde(Recur)
 
 
+@dataclass(unsafe_hash=True)
+class RecurContainer:
+    a: List['RecurContainer']
+    b: Dict[str, 'RecurContainer']
+
+
+serde(Recur)
+
+
 ListPri = List[Pri]
 
 DictPri = Dict[str, Pri]

